### DISCUSSION TITLE
What is the right way to do a search?

### DISCUSSION THREAD

**Discussion by mimimix** [zepp-health/zeppos-forum#241] [Q&A]
I need to search by a set of keywords for a string in an array of 20-30 thousand elements.
For optimization I assigned an id to each word.
An attempt to save the file on the watch was unsuccessful. They can't open a file whose size is more than 50kb +-
Then I decided to try Messaging API, but I didn't find any possibility to read files through it.
The last option was Fetch API, but using it will make the application dependent on the Internet. This is bad.

**Comment by SashaCX75** [comment]
There is an open source application called "Toolbox". On GTR 4 it opens 70kB text files, I just haven't checked more than that. See how it does it, maybe it will help.

**Comment by mimimix** [comment]
I solved my problem by making json files into js files and just importing them.
I know it's not the best, but it's fast.

But the question is still relevant.

**Comment by sirAvent** [comment]
Sounds like an information retrieval problem. Are you searching for keywords that appear in an array of strings? How large are the strings and how many keywords are you looking for? Could be possible for a tf-idf approach.

If it's a substring search, something similar to a Boyer Moore Algorithm can work and will have an Θ(n+m) time complexity.

**Reply by mimimix** [reply]
My app solves chemical reactions
And I just downloaded all the possible :D
A unique identifier has been assigned to each element.
It is necessary to filter reactions by the initial and received elements.
The file has a size of 683KB and consists of such arrays: [[[1,1],[1,1543]],[[1],[5587]]],[[[1,1],[1,119]] 

I'm not working on the app anymore at the moment. But I can share the code
